### PR TITLE
Fix container engine retry logic

### PR DESCRIPTION
**Pull request by mstemm**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

We found some bugs in the container engine lookup retry logic that prevented them from working. These are some fixes.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Fix container lookup retries so they work in the face of temporary failures.
```

### CODE REVIEW

File: userspace/libsinsp/async/async_key_value_source.h [hunk 1]
```diff
@@ -242,6 +244,19 @@ class async_key_value_source
...
+	 * the key back on the request queue with a deadline of now +
+	 * delay ms to allow the run_impl thread to retry the lookup
+	 * later.
+	 *
+	 * If value_ptr is non-NULL, the contents will be saved and provided
+	 * to the next call of dequeue_next_key().
+	 */
+	void defer_lookup(const key_type& key,
+			  value_type* value_ptr,
+			 std::chrono::milliseconds delay);
```
Comment (gnosek) [inline userspace/libsinsp/async/async_key_value_source.h:258]: Indentation seems minimally off?

**Review by gnosek — COMMENTED** [review]
LGTM, though I feel like I'm missing the larger context. What's the issue we're fixing here?

**Comment by mstemm** [comment]
> LGTM, though I feel like I'm missing the larger context. What's the issue we're fixing here?

I'll send you the context offline.

**Review by gnosek — APPROVED** [review]
Your call about the indentation 🤷 :)

**Comment by FedeDP** [comment]
> I'll send you the context offline

Care to share it online too, please :pray: ?

**Comment by mstemm** [comment]
> > LGTM, though I feel like I'm missing the larger context. What's the issue we're fixing here?
> 
> I'll send you the context offline.

Copying the other conversation here for visibility:

It turns out the earlier fix in #688 did fix the problem with returning empty info, but it prevented retries from working. (The existing code wouldn't really work for retries either). You can't really call lookup_delayed() a second time here: https://github.com/falcosecurity/libs/blob/master/userspace/libsinsp/container_engine/container_async_source.tpp#L130, because if there's already an existing request, you don't actually push anything onto the work queue: https://github.com/falcosecurity/libs/blob/master/userspace/libsinsp/async/async_key_value_source.tpp#L173.

From my reading, m_request_queue holds requests that need to be satisfied by the server side, and m_value_map has the values, as a map key -> lookup_request.
lookup_request has the value and some info about the value like whether it's valid yet, a callback to call when the value is ready.
The old contract was when the server side calls dequeue_next_key, it has an obligation to call store_value later.
The new contract is that the server side can either call store_value to provide a value, or defer_lookup to put a request back on the queue with a time in the future.
This allows for server side retries without the server side having to manage its own queue of things to retry.

File: userspace/libsinsp/container_engine/cri.cpp [hunk 2]
```diff
@@ -306,7 +306,8 @@ bool cri::resolve(sinsp_threadinfo *tinfo, bool query_os_for_missing_info)
 
 		if(!m_async_source)
 		{
-			auto async_source = new cri_async_source(cache, m_cri.get(), s_cri_timeout);
+			uint64_t max_wait_ms = 10000;
+			auto async_source = new cri_async_source(cache, m_cri.get(), max_wait_ms);
```
Comment (FedeDP) [inline userspace/libsinsp/container_engine/cri.cpp:310]: Why aren't we using `s_cri_timeout` anymore here?

Comment (mstemm) [reply userspace/libsinsp/container_engine/cri.cpp:310]: I talk about it a little bit in the checkin comments. This param is really about the lifetime of a given value in the async_key_value_source values map and not the timeout for grpc queries when using cri. That's why I switched to its own value.

The cri timeout is still used when doing those grpc queries (see https://github.com/falcosecurity/libs/blob/master/userspace/libsinsp/cri.cpp#L54 and friends)

Comment (FedeDP) [reply userspace/libsinsp/container_engine/cri.cpp:310]: Thanks; makes sense!

**Review by FedeDP — APPROVED** [review]
/approve

**Pull request merged**